Request cannot span more than 24 hours Error

I am starting to see odd errors in our logs from CloudWatch Log node when running the application live.

```
[ WARN] [1593087736.376730661, 30.901000000]: [AWSErrorMarshaller] Encountered AWSError 'InvalidParameterException': T
he batch of log events in a single PutLogEvents request cannot span more than 24 hours.
[ERROR] [1593087736.376854123, 30.901000000]: [AWSClient] HTTP response code: 400
Exception name: InvalidParameterException
Error message: The batch of log events in a single PutLogEvents request cannot span more than 24 hours.
5 response headers:
connection : close
content-length : 139
content-type : application/x-amz-json-1.1
date : Thu, 25 Jun 2020 12:22:15 GMT
x-amzn-requestid : 6e2c66d4-baf8-48b8-af82-bb56bad020f0
[ WARN] [1593087736.376925890, 30.901000000]: [AWSClient] If the signature check failed. This could be because of a ti
me skew. Attempting to adjust the signer.
[ERROR] [1593087736.377026834, 30.901000000]: [SendLogsRequest] Send log request failed due to: The batch of log event
s in a single PutLogEvents request cannot span more than 24 hours., with error code: 131
[ERROR] [1593087736.377057525, 30.901000000]: [SendLogsToCloudWatch] Failed to send to CloudWatch in Log Group: ab200s
im Log Stream: 4a54bac8-efb7-47a4-9ee5-493d928ae446 with error code: 1
```

What is really odd is after this error we still get application shipped to CloudWatch and have been using them for weeks. 

Things I have checked:
- The timestamp is consistent with the epoch time, no skewed system time
- Latest updates for the package

Using `Version: 2.3.1-1xenial-20200330-153719+0000`

TIL that CWL doesn't except logs older than 2 weeks as well. I think for our application we are OK with that. The concern I have is  error on the the log node seems to be when uploading logs that are within the 2 week period.

For example these files are within the last 4ish days
```
$ ls ~/.ros/cwlogs 
cwlog2020-06-23_09-10-58-0.log  cwlog2020-06-25_08-42-32-0.log  token_store.info
cwlog2020-06-23_09-16-23-0.log  cwlog2020-06-26_18-15-05-0.log
```


And still receive the same `PutLogEvents request cannot span more than 24 hours., with error code: 13` error.


```
[ INFO] [1593213347.383233723, 30.229000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1593213347.537750453, 30.311000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1593213347.688859787, 30.428000000]: [SendLogs] Sending logs to CW
[ INFO] [1593213347.689853891, 30.428000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1593213347.973261900, 30.645000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1593213347.973564004, 30.645000000]: [readBatch] Reading Logbatch
[ INFO] [1593213347.978477776, 30.647000000]: [readBatch] Reading Logbatch
[ INFO] [1593213347.979023272, 30.647000000]: [SendLogs] Sending logs to CW
[ INFO] [1593213347.979830340, 30.647000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1593213347.981544168, 30.647000000]: [readBatch] Reading Logbatch
[ INFO] [1593213347.992540090, 30.648000000]: [read] Data is no longer available to read.
[ INFO] [1593213348.185021522, 30.768000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1593213348.185087880, 30.768000000]: [fileUploadCompleteStatus] Total items uploaded: 50
[ INFO] [1593213348.185470798, 30.768000000]: [SendLogs] Sending logs to CW
[ INFO] [1593213348.187257322, 30.768000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ WARN] [1593213348.405147533, 30.961000000]: [AWSErrorMarshaller] Encountered AWSError 'InvalidParameterException': T
he batch of log events in a single PutLogEvents request cannot span more than 24 hours.
[ERROR] [1593213348.405273183, 30.961000000]: [AWSClient] HTTP response code: 400
Exception name: InvalidParameterException
Error message: The batch of log events in a single PutLogEvents request cannot span more than 24 hours.
5 response headers:
connection : close
content-length : 139
content-type : application/x-amz-json-1.1
date : Fri, 26 Jun 2020 23:15:47 GMT
x-amzn-requestid : fb38f16b-1445-4ef1-a9c7-f6b21ab9de2d
[ WARN] [1593213348.405358547, 30.961000000]: [AWSClient] If the signature check failed. This could be because of a ti
me skew. Attempting to adjust the signer.
[ERROR] [1593213348.406029900, 30.962000000]: [SendLogsRequest] Send log request failed due to: The batch of log event
s in a single PutLogEvents request cannot span more than 24 hours., with error code: 131
[ERROR] [1593213348.406071218, 30.962000000]: [SendLogsToCloudWatch] Failed to send to CloudWatch in Log Group: ab200s
im Log Stream: 4a54bac8-efb7-47a4-9ee5-493d928ae446 with error code: 1
[ WARN] [1593213348.406098423, 30.962000000]: [SendLogs] Unable to send logs to CloudWatch, retrying ...
[ INFO] [1593213348.406315606, 30.962000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ WARN] [1593213348.843280890, 31.312000000]: [SendLogs] Unable to send logs to CloudWatch
[ WARN] [1593213348.843455163, 31.312000000]: [publishData] error finishing SendLogs 1
[ INFO] [1593213348.843488525, 31.312000000]: [onPublisherStateChange] Publisher state has changed to: unavailable
```

All those logs are within the 2 week and should be able to be uploaded.

I have been doing some testing to try to find a good set of parameters and so far no luck.

Here is what I have tried:
`file_max_queue_size` : 1, this makes sense as each file in `~/.ros/cwlogs` should be a 24 hour period?
`file_upload_batch_size`:  1-10 (5 in the below log), at first glance seems to work but will hit a throttle error

The sending of cached logs does not seem to honor the publish frequency and then runs into the CW Log API throttle.
```
[ INFO] [1594123577.100182436, 34.639000000]: [fileUploadCompleteStatus] Total items uploaded: 215
[ INFO] [1594123577.101845707, 34.639000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123577.102641180, 34.643000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1594123577.115053568, 34.659000000]: [readBatch] Reading Logbatch
[ INFO] [1594123577.279790079, 34.768000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1594123577.280345720, 34.768000000]: [fileUploadCompleteStatus] Total items uploaded: 220
[ INFO] [1594123577.280768058, 34.768000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123577.281258849, 34.768000000]: [readBatch] Reading Logbatch
[ INFO] [1594123577.282001411, 34.768000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1594123577.449216726, 34.842000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1594123577.449806603, 34.842000000]: [fileUploadCompleteStatus] Total items uploaded: 225
[ INFO] [1594123577.450003619, 34.842000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123577.450735318, 34.842000000]: [readBatch] Reading Logbatch
[ INFO] [1594123577.451164743, 34.842000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1594123577.616988298, 34.905000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1594123577.617050819, 34.905000000]: [fileUploadCompleteStatus] Total items uploaded: 230
[ INFO] [1594123577.617249252, 34.905000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123577.617523011, 34.905000000]: [readBatch] Reading Logbatch
[ INFO] [1594123577.618274838, 34.905000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1594123577.784455858, 34.991000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1594123577.784533900, 34.991000000]: [fileUploadCompleteStatus] Total items uploaded: 235
[ INFO] [1594123577.784772493, 34.991000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123577.785212344, 34.991000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ INFO] [1594123577.785425092, 34.991000000]: [readBatch] Reading Logbatch
[ INFO] [1594123577.950393045, 35.102000000]: [onPublisherStateChange] Publisher state has changed to: available
[ INFO] [1594123577.951176582, 35.102000000]: [fileUploadCompleteStatus] Total items uploaded: 240
.
.
.
[ WARN] [1594123581.721042767, 37.140000000]: [AWSErrorMarshaller] Encountered AWSError 'ThrottlingException': Rate exceeded
[ERROR] [1594123581.721934338, 37.140000000]: [AWSClient] HTTP response code: 400
Exception name: ThrottlingException
Error message: Rate exceeded
5 response headers:
connection : close
content-length : 58
content-type : application/x-amz-json-1.1
date : Tue, 07 Jul 2020 12:06:21 GMT
x-amzn-requestid : 70ebd053-3a73-4114-890d-1f3e81dc0eda
```

Setting anything larger than 10 I have hit a 24 hour period issue as before which is odd.  Here is a test with a logs file a week old.
```
 $ ls ~/.ros/cwlogs                  
cwlog2020-07-01_09-37-37-0.log  cwlog2020-07-07_07-05-07-0.log
```

The ROS log of it failing on 24hour when `file_upload_batch_size` : 50 (default)
```
[ INFO] [1594123961.172877999, 31.332000000]: [fileUploadCompleteStatus] Total items uploaded: 350
[ INFO] [1594123961.203803237, 31.336000000]: [SendLogs] Sending logs to CW
[ INFO] [1594123961.206706029, 31.336000000]: [CurlHandleContainer] Connection has been released. Continuing.
[ WARN] [1594123961.410863358, 31.363000000]: [AWSErrorMarshaller] Encountered AWSError 'InvalidParameterException': T
he batch of log events in a single PutLogEvents request cannot span more than 24 hours.
[ERROR] [1594123961.411046135, 31.363000000]: [AWSClient] HTTP response code: 400
Exception name: InvalidParameterException
Error message: The batch of log events in a single PutLogEvents request cannot span more than 24 hours.
5 response headers:
connection : close
content-length : 139
content-type : application/x-amz-json-1.1
date : Tue, 07 Jul 2020 12:12:41 GMT
x-amzn-requestid : f1bf30fd-3ebd-46d9-8d1e-17a3e32145a9
```

My guess is the `file_max_queue_size` is reading more than a single file thus hitting the 24 hour span error. Thoughts?

--- a/README.md
+++ b/README.md
@@ -124,7 +124,7 @@ An example configuration file called `sample_configuration.yaml` is provided. Wh
 | publish_topic_names | Whether or not to include topic name information in the log messsages that are uploaded to AWS CloudWatch Logs | *bool* | true/false | true |
 | storage_directory | The location where all offline metrics will be stored | *string* | string | ~/.ros/cwlogs/ |
 | storage_limit | The maximum size of all offline storage files in KB. Once this limit is reached offline logs will start to be deleted oldest first. | *int* | number | 1048576 |
-| delete_stale_data | Whether or not to delete log batch data that are over 14 days old, which are rejected by AWS PutLogEvents. | *bool* | true/false | false |
+| delete_stale_data | Whether or not to delete log batch data that are over 14 days old, which are rejected by [AWS PutLogEvents](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html). | *bool* | true/false | false |
 | aws_client_configuration | AWS region configuration | *std::string* | *region*: "us-west-2"/"us-east-1"/"us-east-2"/etc. | region: us-west-2 |
 
 ### Advanced Configuration Parameters

--- a/README.md
+++ b/README.md
@@ -124,6 +124,7 @@ An example configuration file called `sample_configuration.yaml` is provided. Wh
 | publish_topic_names | Whether or not to include topic name information in the log messsages that are uploaded to AWS CloudWatch Logs | *bool* | true/false | true |
 | storage_directory | The location where all offline metrics will be stored | *string* | string | ~/.ros/cwlogs/ |
 | storage_limit | The maximum size of all offline storage files in KB. Once this limit is reached offline logs will start to be deleted oldest first. | *int* | number | 1048576 |
+| delete_stale_data | Whether or not to delete log batch data that are over 14 days old, which are rejected by AWS PutLogEvents. | *bool* | true/false | false |
 | aws_client_configuration | AWS region configuration | *std::string* | *region*: "us-west-2"/"us-east-1"/"us-east-2"/etc. | region: us-west-2 |
 
 ### Advanced Configuration Parameters

--- a/cloudwatch_logger/config/sample_configuration.yaml
+++ b/cloudwatch_logger/config/sample_configuration.yaml
@@ -44,6 +44,10 @@ storage_directory: "~/.ros/cwlogs/"
 # The maximum size of all files in offline storage in KB
 storage_limit: 1048576
 
+# whether or not to automatically delete log data > 14 days old
+#default value is: false
+delete_stale_data: false
+
 # This is the AWS Client Configuration used by the AWS service client in the Node. If given the node will load the
 # provided configuration when initializing the client.
 aws_client_configuration:

--- a/cloudwatch_logger/include/cloudwatch_logger/log_node_param_helper.h
+++ b/cloudwatch_logger/include/cloudwatch_logger/log_node_param_helper.h
@@ -232,6 +232,22 @@ void ReadOption(
   const size_t & default_value,
   size_t & option_value);
 
+/**
+ * Fetch a single size_t option
+ *
+ * @param parameter_reader to retrieve the parameters from
+ * @param option_key the parameter key to read
+ * @param default_value a default value if the parameter doesn't exist or is unreadble
+ * @param option_value the size_t value for this option
+ * @return an error code that indicates whether the parameter was read successfully or not,
+ * as returned by \p parameter_reader
+ */
+void ReadOption(
+  const std::shared_ptr<Aws::Client::ParameterReaderInterface>& parameter_reader,
+  const std::string & option_key,
+  const bool & default_value,
+  bool & option_value);
+
 }  // namespace Utils
 }  // namespace CloudWatchLogs
 }  // namespace Aws

--- a/cloudwatch_logger/include/cloudwatch_logger/log_node_param_helper.h
+++ b/cloudwatch_logger/include/cloudwatch_logger/log_node_param_helper.h
@@ -51,6 +51,7 @@ constexpr char kNodeParamStorageDirectory[] = "storage_directory";
 constexpr char kNodeParamFileExtension[] = "file_extension";
 constexpr char kNodeParamMaximumFileSize[] = "maximum_file_size";
 constexpr char kNodeParamStorageLimit[] = "storage_limit";
+constexpr char kNodeParamDeleteStaleData[] = "delete_stale_data";
 
 /** Default values for parameters **/
 constexpr char kNodeLogGroupNameDefaultValue[] = "ros_log_group";

--- a/cloudwatch_logger/src/log_node_param_helper.cpp
+++ b/cloudwatch_logger/src/log_node_param_helper.cpp
@@ -389,6 +389,29 @@ void ReadOption(
   }
 }
 
+void ReadOption(
+  const std::shared_ptr<Aws::Client::ParameterReaderInterface>& parameter_reader,
+  const std::string & option_key,
+  const bool & default_value,
+  bool & option_value)
+{
+  Aws::AwsError ret = parameter_reader->ReadParam(ParameterPath(option_key), option_value);
+  switch (ret) {
+    case Aws::AwsError::AWS_ERR_NOT_FOUND:
+      option_value = default_value;
+      AWS_LOGSTREAM_INFO(__func__,
+                         option_key << " parameter not found, setting to default value: " << default_value);
+      break;
+    case Aws::AwsError::AWS_ERR_OK:
+      AWS_LOGSTREAM_INFO(__func__, option_key << " is set to: " << option_value);
+      break;
+    default:
+      option_value = default_value;
+      AWS_LOGSTREAM_ERROR(__func__,
+                          "Error " << ret << " retrieving option " << option_key << ", setting to default value: " << default_value);
+  }
+}
+
 }  // namespace Utils
 }  // namespace CloudWatchLogs
 }  // namespace Aws

--- a/cloudwatch_logger/src/log_node_param_helper.cpp
+++ b/cloudwatch_logger/src/log_node_param_helper.cpp
@@ -335,6 +335,12 @@ void ReadFileManagerStrategyOptions(
     kNodeParamStorageLimit,
     Aws::FileManagement::kDefaultFileManagerStrategyOptions.storage_limit_in_kb,
     file_manager_strategy_options.storage_limit_in_kb);
+
+  ReadOption(
+    parameter_reader,
+    kNodeParamDeleteStaleData,
+    Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data,
+    file_manager_strategy_options.delete_stale_data);
 }
 
 void ReadOption(

--- a/cloudwatch_logger/test/log_node_param_helper_test.cpp
+++ b/cloudwatch_logger/test/log_node_param_helper_test.cpp
@@ -319,6 +319,86 @@ TEST_F(LogNodeParamHelperFixture, TestReadIgnoreNodesSet)
     EXPECT_EQ(AwsError::AWS_ERR_OK, ReadIgnoreNodesSet(param_reader_, param));
     EXPECT_EQ(1, param.count("String1"));
 }
+/**
+ * FileManagerStrategyOptions options defined with delete_stale_data set to true.
+ * First ReadParam call fails, therefore we expect options_.delete_stale_data set true->false.
+ * Second ReadParam call suceeds, therefore we expect options_.delete_stale_data set false->true.
+ * Third ReadParam call not found, therefore we expect options_.delete_stale_data set true->false.
+ */
+TEST_F(LogNodeParamHelperFixture, Test_Delete_Stale_Data_True)
+{
+  {
+    InSequence read_param_seq;
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(Return(AwsError::AWS_ERR_FAILURE));
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(
+      DoAll(SetArgReferee<1>(true), Return(AwsError::AWS_ERR_OK))
+    );
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(Return(AwsError::AWS_ERR_NOT_FOUND));
+  }
+
+  Aws::FileManagement::FileManagerStrategyOptions options_{"test", "log_tests/", ".log", 1024*1024, 1024*1024, true};
+  ASSERT_TRUE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_FALSE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_TRUE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_FALSE(options_.delete_stale_data);
+}
+/**
+ * FileManagerStrategyOptions options defined with delete_stale_data not defined.
+ * We expect that when delete_stale_data is not defined that it will default to false.
+ * First ReadParam call suceeds, therefore we expect options_.delete_stale_data set false->true.
+ * Second ReadParam call fails, therefore we expect options_.delete_stale_data set true->false.
+ * Third ReadParam call suceeds, therefore we expect options_.delete_stale_data set false->true.
+ * Forth ReadParam call not found, therefore we expect options_.delete_stale_data set true->false.
+ */
+TEST_F(LogNodeParamHelperFixture, Test_Delete_Stale_Data_Not_Defined)
+{
+  {
+    InSequence read_param_seq;
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(
+      DoAll(SetArgReferee<1>(true), Return(AwsError::AWS_ERR_OK))
+    );
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(Return(AwsError::AWS_ERR_FAILURE));
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(
+      DoAll(SetArgReferee<1>(true), Return(AwsError::AWS_ERR_OK))
+    );
+
+    EXPECT_CALL(*param_reader_, ReadParam(Eq(ParameterPath(kNodeParamDeleteStaleData)), A<bool &>()))
+    .WillOnce(Return(AwsError::AWS_ERR_NOT_FOUND));
+  }
+
+  Aws::FileManagement::FileManagerStrategyOptions options_{"test", "log_tests/", ".log", 1024*1024, 1024*1024};
+  ASSERT_FALSE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_TRUE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_FALSE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_TRUE(options_.delete_stale_data);
+
+  ReadOption(param_reader_, kNodeParamDeleteStaleData, Aws::FileManagement::kDefaultFileManagerStrategyOptions.delete_stale_data, options_.delete_stale_data);
+  ASSERT_FALSE(options_.delete_stale_data);
+}
 
 int main(int argc, char ** argv)
 {

